Add type hints support (#355)

JoshCLWren opened the pull request:
Closes #355

This PR adds PEP 561 type stubs for pyvips, enabling IDE autocomplete and type checking with mypy.

## Changes
- Add `pyvips/__init__.pyi` - Type stubs for all Image operations (300+ methods)
- Add `pyvips/generate_type_stubs.py` - Script to regenerate stubs from libvips introspection
- Add `tests/test_type_hints.py` - Test verifying type stubs work correctly
- Update README.rst with type checking section
- Add mypy check to CI workflow

## Design Decisions
Type stubs are **generated** rather than handwritten because:
- pyvips is a binding around libvips C library, which evolves independently
- libvips has 300+ operations with complex signatures
- New operations are added to libvips frequently
- Generation uses same introspection as existing doc/enum generation
- Ensures stubs stay synchronized with available operations

This follows the existing pattern used for:
- Enum generation (`examples/gen-enums.py`)
- Documentation generation (`pyvips.Operation.generate_sphinx_all()`)

## Features
- ✅ Complete coverage of Image methods via introspection
- ✅ No use of `Any` - specific Union types instead
- ✅ Supports both string and enum values for enum parameters
- ✅ Python keyword escaping (e.g., `in_` for `in` parameter)
- ✅ Proper typing for optional outputs with dictionaries
- ✅ CI integration to validate stubs
- ✅ Zero runtime overhead (stubs only used by type checkers)

## Usage
Users can now type check pyvips code:

```bash
pip install mypy
mypy your_script.py
```

Regenerate stubs after libvips updates:
```bash
python pyvips/generate_type_stubs.py
```

## Testing
- All existing tests pass (48 passed, 7 skipped)
- New `test_type_hints.py` added
- Mypy validation passes

JoshCLWren committed:
Add type hints support (#355)

Adds PEP 561 type stubs for pyvips, enabling IDE autocomplete
and type checking with mypy.

- Add pyvips/__init__.pyi with type stubs for 300+ Image operations
- Add generate_type_stubs.py script using introspection (follows existing enum/doc pattern)
- Add test_type_hints.py with mypy validation
- Update README with type checking documentation
- Add mypy check to CI workflow

Type stubs are generated to handle libvips's 300+ operations
and frequent updates. Zero runtime overhead, full type coverage.

jcupitt commented:
This is great @JoshCLWren!

I'm on holiday right now, but I'll read this carefully when I get home again.

jcupitt commented on pyvips/generate_type_stubs.py:
```diff
@@ -0,0 +1,456 @@
+    stub += """
+
+    # Operator overloads
+    def __add__(self, other: Union[Image, float, int]) -> Image: ...
```
You can have arrays of float and int too, eg. `a + [1, 2, 3]` adds 1 to band 0, 2 to band 1, 3 to band 2.

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,456 @@
+    def __call__(self, x: int, y: int) -> List[float]: ...
+    def __repr__(self) -> str: ...
+
+
```
There are a set of hand-written bindings too, perhaps some of them could have type hints? 

Easy: `image.floor()` etc.

Medium: `bandjoin`, `bandsplit`, `composite` etc.

Horrible: `ifthenelse`.

jcupitt reviewed (commented):
Other:

- we need a line in the changelog, and please credit yourself for doing this nice work
- `generate_type_stubs.py` needs to be set executable
- does the type stub generator need to be in `pyvips/`? it's something for maintainers, so I'd be inclined to put it in `examples/`. The doc generator is in `pyvips/` because it's used for `help()` output
- perhaps we could run mypy on the demo scripts in `examples/` as part of CI? but it'll need eg. a hint for `image * [1, 2, 3]` first

JoshCLWren committed:
Add comprehensive type hints for hand-written bindings and enhance operator overloads

- Add type hints for ifthenelse, composite, floor, ceil, rint, bandsplit, bandjoin, bandrank, hasalpha, get_page_height
- Enhance operator overloads to support List[float] and List[int] operands
- Move generate_type_stubs.py from pyvips/ to examples/ for better maintainability
- Add CI mypy checks for examples/affine.py and examples/convolve.py
- Update documentation references for type stub generation path

JoshCLWren committed:
Update CHANGELOG credit to JoshCLWren

JoshCLWren committed:
Add type hints for composite and ifthenelse hand-written bindings

JoshCLWren committed:
Expand mypy type checking to more example scripts

jcupitt commented on pyvips/__init__.pyi:
```diff
@@ -87,23 +86,50 @@ class Image:
     @staticmethod
-    def new_from_list(array: List[List[float]], scale: float =1.0, offset: float = 0.0) -> Image: ...
+    def new_from_list(
+        array: List[List[float]], scale: float = 1.0, offset: float = 0.0
```
I think this can also be `List[List[int]]`

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+    enum_classes = "\n".join([f"class {name}: ..." for name in enum_names])
+
+    stub = f'''"""Type stubs for pyvips.
+
```
Add:

```python
    # flake8: noqa: E501
```

to stop flake8 moaning about long lines in the generated output.

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+"""
+
+import typing
+from typing import Any, Optional, Union
```
You don't use any of these (I think).

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+    type_map,
+    type_from_name,
+    nickname_find,
+    at_least_libvips,
```
unused

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+    nickname_find,
+    at_least_libvips,
+)
+from pyvips import ffi, Error, _to_bytes, _to_string
```
`_to_byes`, `_to_string` unused

jcupitt commented on tests/test_type_hints.py:
```diff
@@ -0,0 +1,38 @@
+    assert isinstance(typeof_val, int)
+
+    # Test get returns union type
+    metadata = img.get("xres")
```
needs an assert

jcupitt commented on examples/watermark.py:
```diff
@@ -5,23 +5,27 @@
 
 # scale the alpha down to make the text semi-transparent
-text = (text * [1, 1, 1, 0.3]).cast("uchar")
+text = (text * [1, 1, 1, 0.3]).cast("uchar")  # type: ignore[operator]
```
do you need the ignore now that you have a hint for image `*` array?

jcupitt commented on CHANGELOG.rst:
```diff
@@ -3,6 +3,7 @@
 - fix get_gainmap arguments [jcupitt]
 - fix __array__ copy argument test [jcupitt]
 - add a test [jcupitt]
+- add type stub generation script with support for all libvips operations [JoshCLWren]
```
We'll need a new version -- let's make this 3.2 and move this changelog line to that group.

jcupitt commented on examples/try5.py:
```diff
@@ -23,15 +23,13 @@ def bandsplit(a):
 
 b = a + [12, 0, 0]
-x = map(lambda x: x.avg())
-bandsplit(a)
```
Haha that's been broken for a while.

jcupitt commented on examples/try5.py:
```diff
@@ -48,7 +46,7 @@ def bandsplit(a):
 b = a * b
 b = 12 * a
-b = [12, 1, 1] * a
+b = [12, 1, 1] * a  # type: ignore[assignment, operator]
```
is the ignore needed?

jcupitt commented on examples/try5.py:
```diff
@@ -66,11 +64,11 @@ def bandsplit(a):
 b = a % b
 
-b = a ** 12
+b = a**12
```
I don't like this -- we use spaces around `**` everywhere else, we should use spaces here too.

jcupitt commented on examples/watermark.py:
```diff
@@ -5,23 +5,27 @@
+    dpi=100,
+    align="centre",
+    rgba=True,
+)
```
I don't like this change, we indent everywhere else on the bracket, we should stay consistent.

jcupitt commented on examples/watermark.py:
```diff
@@ -5,23 +5,27 @@
 page_height = im.get_page_height()
-text = text.replicate(1 + im.width / text.width, 1 + page_height / text.height)
+text = text.replicate(
+    int(1 + im.width / text.width), int(1 + page_height / text.height)
```
same here

jcupitt commented on README.rst:
```diff
@@ -228,6 +228,30 @@ Stylecheck:
+    $ mypy your_script.py
+
+Note: ``pyvips`` methods accept arbitrary keyword arguments for libvips options,
+which may not be fully covered by type hints.
```
Is this correct? We seem to have them all covered, I think.

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+    import os
+
+    # Script is in examples/, need to go up one level to find pyvips/
+    stub_file = os.path.join(os.path.dirname(__file__), "..", "pyvips", "__init__.pyi")
```
I'm not sure we need this -- just generate `__init__.pyi` in the current directory and let the user know. Or just print to stdout (the safest).

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+    """Generate type signature for an operation method."""
+    intro = Introspect.get(operation_name)
+
+    if (intro.flags & 4) != 0:  # _OPERATION_DEPRECATED
```
Is this correct? `VipsOperationFlags` uses 8 for `VipsOperationFlags`. I think you're removed all uncacheable operations instead.

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+
+    # Return type
+    output_types = [
+        gtype_to_python_type(intro.details[name]["type"])
```
no newline before this

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+
+    # Optional output dicts can contain any metadata value type
+    if len(intro.doc_optional_output) > 0:
+        dict_value_type = (
```
We don't need the brackets

jcupitt commented on examples/generate_type_stubs.py:
```diff
@@ -0,0 +1,469 @@
+    return "\n".join([sig for _, sig in all_names])
+
+
+def get_all_enum_names() -> list[str]:
```
I suppose we should handle flags as well as enums, but perhaps that can wait for another PR.

jcupitt replied:
gen-enums has this code as well, of course, perhaps it should be moved into `Introspect`

jcupitt commented:
Sorry, there are a lot of comments. I don't know how much time you have -- we could merge and share the work of fixups in later PRs, if you like.

jcupitt commented:
Oh, I just saw cdisplayagain, nice!

You have:

> The project requires pyvips for fast image processing. Install the libvips library via your package manager:

Another possible solution is:

```console
$ pip install pyvips[binary]
```

And it should download and use a libvips binary appropriate for your system. It might be a little easier for your users.

JoshCLWren commented:
> Sorry, there are a lot of comments. I don't know how much time you have -- we could merge and share the work of fixups in later PRs, if you like.

No worries! Thanks for the feedback. I'll get on it this afternoon/evening.

JoshCLWren commented:
> Oh, I just saw cdisplayagain, nice!
> 
> You have:
> 
> > The project requires pyvips for fast image processing. Install the libvips library via your package manager:
> 
> Another possible solution is:
> 
> ```
> $ pip install pyvips[binary]
> ```
> 
> And it should download and use a libvips binary appropriate for your system. It might be a little easier for your users.

Good call, I think that'll help a lot.

JoshCLWren committed:
Fix operation filtering in type stub generator to correctly filter deprecated operations

Changed hardcoded value 4 (NOCACHE) to use _OPERATION_DEPRECATED constant (8).
This fixes incorrect filtering that was excluding uncacheable operations instead
of deprecated ones.

JoshCLWren committed:
Clean up type stub generator: remove unused imports, extra blank line, add flake8 noqa

JoshCLWren committed:
Complete comprehensive type hints for pyvips

- Add SourceCustom class with on_read() and on_seek() methods
- Add Source.new_from_descriptor() and Target.new_to_descriptor() static methods
- Add pyvips.call() function for calling libvips operations
- Add erode() and dilate() methods to Image class stub
- Add explicit enum attributes for Direction and Align (HORIZONTAL, VERTICAL, LOW, CENTRE, HIGH)
- Fix try7.py to use .format instead of non-existent .bandfmt attribute
- Add type: ignore comments for Union type narrowing issues in 7 example files
- All 35 example files now pass mypy type checking
- Type stub file passes mypy with 0 errors

JoshCLWren committed:
Update version to 3.2.0 and update CI

- Update version.py to 3.2.0
- Update CHANGELOG.rst with version 3.2.0 entry
- Update doc/conf.py to version 3.2.0
- Expand CI mypy checks to include more example scripts

JoshCLWren committed:
Fix new_from_image type annotation to include int and List[int] support

Match operator overload signatures which already support int and List[int] types.
Ensures type consistency across the Image API.

JoshCLWren committed:
Revert linting changes in examples and doc/conf.py

- Revert examples/*.py to origin/master (remove type: ignore comments that caused flake8 E501 errors)
- Revert doc/conf.py quote-style changes back to single quotes
- Keep only version bump (3.1 -> 3.2, 3.1.1 -> 3.2.0)
- Fix remaining line-length issues in doc/conf.py

JoshCLWren committed:
Fix doc/conf.py syntax and line-length errors

JoshCLWren committed:
pin mypy to <1.19 for PyPy 3.9 compatibility

mypy 1.19+ introduced a dependency on the librt module which is not
available on PyPy 3.9, causing ModuleNotFoundError during type checking.
This pins mypy to version 1.18.x which works across all Python versions
in the CI matrix.

jcupitt commented:
Sorry for the pause. Let's merge!

Thank you again for doing this work, @JoshCLWren.
